embassy-stm32/src/spi/mod.rs

```diff
@@ -117,6 +151,7 @@ impl<'d, T: Instance, Tx, Rx> Spi<'d, T, Tx, Rx> {
         )
     }
 
+    // TODO: it receive only on master, on slave in miso is expecting to write something
```

Comment:
What do you think if we will refactor api to make it builder like? It could be second way for initialization.

```rust
let builder = SpiBuilder::new(p.SPI3)
  .sck( p.PC10)
  .mosi(p.PC12)
  .build(Config { // runtime config can be kept in build, or moved to other build methods.
      frequency: Hz(1_000_000),
      ..Default::default()
});
```
vs:
```rust
let mut spi = Spi::new(
    p.SPI3,
    p.PC10,
    p.PC12,
    p.PC11,
    NoDma,
    NoDma,
    Hertz(1_000_000),
    Config::default(),
);

```

https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=cd53ccfdc02245cadebf0fa07a70b35d

Reply:
We're intentionally not using builders, because they become very complicated when they interact with generics. (for example the fact that calling `.mosi()` changes the `Mosi` generic param). Also, we had a `NoPin` exactly like yours in the past, it was removed, see #605.

Multiple constructors is the most straightforward way to do it, and there's little code duplication because they internally all call into `new_inner`.

Reply:
Yes, i usually don't like builders too, but for pin configuration (because they usually are static typed) i found it verbose and clear. 

The problem with constructors is that they params is annonymous, and it is much easier to make a mess with pin configs.

Reply:
I agree having so many params is unfortunate. However, modern IDEs can show inline hints with the param names (this pic is from rust-analyzer + vscode with the default config), so I don't think this justifies having a builder.

![Uploading screenshot-2022-08-01_00-44-33.png…]()